crates/bevy_ecs/src/component.rs

```diff
@@ -93,6 +97,91 @@ use std::{
 /// [`Table`]: crate::storage::Table
 /// [`SparseSet`]: crate::storage::SparseSet
 ///
+/// # Required Components
+///
+/// Components can specify Required Components. If some [`Component`] `A` requires [`Component`] `B`,  then when `A` is inserted,
+/// `B` will _also_ be initialized and inserted (if it was not manually specified).
+///
+/// The [`Default`] constructor will be used to initialize the component, by default:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #[derive(Component)]
+/// #[require(B)]
+/// struct A;
+///
+/// #[derive(Component, Default)]
+/// struct B(usize);
+///
+/// # let mut world = World::default();
+/// // This will implicitly also insert B with the Default constructor
+/// world.spawn(A);
+///
+/// // This will _not_ implicitly insert B, because it was already provided
+/// world.spawn((A, B(11)));
+/// ```
+///
+/// Components can have more than one required component:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #[derive(Component)]
+/// #[require(B, C)]
+/// struct A;
+///
+/// #[derive(Component, Default)]
+/// #[require(C)]
+/// struct B(usize);
+///
+/// #[derive(Component, Default)]
+/// struct C(f32);
+///
+/// # let mut world = World::default();
+/// // This will implicitly also insert B and C with their Default constructors
+/// world.spawn(A);
+/// ```
+///
+/// You can also define a custom constructor:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #[derive(Component)]
+/// #[require(B(init_b))]
+/// struct A;
+///
+/// #[derive(Component)]
+/// struct B(usize);
+///
+/// fn init_b() -> B {
+///     B(10)
+/// }
+///
+/// # let mut world = World::default();
+/// // This will implicitly also insert B with the init_b() constructor
+/// world.spawn(A);
+/// ```
+///
+/// Required components are _recursive_. This means, if a Required Component has required components,
+/// those components will _also_ be inserted if they are missing:
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// #[derive(Component)]
+/// #[require(B)]
+/// struct A;
+///
+/// #[derive(Component, Default)]
+/// #[require(C)]
+/// struct B(usize);
+///
+/// #[derive(Component, Default)]
+/// struct C(f32);
+///
+/// # let mut world = World::default();
+/// // This will implicitly also insert B and C with their Default constructors
+/// world.spawn(A);
+/// ```
+///
```

Comment:
What happens in the following scenario?
```rs
#[derive(Component, Default)]
#[require(B<usize>)]
struct B<T>(T);

world.spawn(B(1_u8));
```
Does it create `(B(1_u8), B(0_usize))` or does `B(0_usize)` try require itself & do something weird?

Reply:
This would result in a stack overflow at insert time, as `B<usize>` would also require `B<usize>`.